ql/termstructures/globalbootstrap.hpp

```diff
@@ -304,7 +298,8 @@ template <class Curve> void GlobalBootstrap<Curve>::calculate() const {
         Curve *ts_;
         const std::vector<Real> lowerBounds_, upperBounds_;
```

Comment:
This class is local to the function, instantiated and then destroyed at function end.  You can declare `lowerBounds_` and `upperBounds_` as `const &` instead and avoid all the moves.

Reply:
Sure, but note that constructor arguments for `TargetFunction` then should *not* be declared as `const&`, because doing so would allow passing temporary objects. We are then storing a reference to a temporary object, which will be destroyed at the end of the statement. I declared them as const pointers instead.

Reply:
It's safe at the moment, but we might refractor the code in the future to, say, use std ranges. I'd rather have a compile time error than debug random numbers coming out from bootstrapping.

Reply:
My two cents is that const reference is better syntactically and stylistically, and that any regression from future refactoring will be caught right away by the address sanitizer CI job as long as there are adequate unit tests in place.

Reply:
Please wrote libraries to avoid this exact issue: https://github.com/akrzemi1/explicit/blob/master/doc/lvalue_ref.md
So no, it's not better.

Reply:
While QuantLib doesn't strictly enforce a specific style, I personally like the Google C++ Style Guide, and after rechecking it now I see that it suggests the [following](https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs:~:text=pass%20retained%20parameters%20by%20pointer%20and%20document%20the%20lifetime%20and%20non%2Dnull%20requirements.):

> pass retained parameters by pointer and document the lifetime and non-null requirements.

So I take back what I said about const reference being better style. I would recommend adding a comment as per the guide and maybe also a `QL_REQUIRE` to check for null values being passed in.

Reply:
Yes, Google C++ style prohibits non-const references and requires using pointers to mutate arguments, so using pointers for retained arguments is very natural.

Reply:
Too bad the interface of Problem requires this to inherit from CostFunction.  Otherwise, the whole thing could be a lambda and just capture what it needs.  Oh well.

Reply:
Good point. We can make a templated implementation of CostFunction that wraps a functor. Then we can write this as a lambda.

Reply:
How's this https://github.com/lballabio/QuantLib/pull/2117?

Reply:
Great, thanks.